pkg/token/service.go

```diff
@@ -423,72 +437,104 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
-			return fmt.Errorf("unable to save access token: %w", err)
-		}
-		return nil
-	})
-	if err != nil {
-		return err
+	for _, k := range d.rtTargetDomainRolesToFile {
+		wg.Add(1)
+		go func(k CacheKey) {
+			defer wg.Done()
+			domain, role := k.Domain, k.Role
+			token := d.roleTokenCache.Load(k)
+			outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
+			err := d.writeFileWithRetry(ctx, maxElapsedTime, token, outPath, mROLE_TOKEN)
+			if err != nil {
+				echan <- err
+				rtErrorCount.Add(1)
+			}
+		}(k)
 	}
-	err = d.roleTokenCache.Range(func(k CacheKey, t Token) error {
-		domain := t.Domain()
-		role := t.Role()
-		rt := t.Raw()
-		if rt == "" {
-			// skip placeholder token added during daemon creation
-			return nil
-		}
 
-		log.Infof("[New Role Token] Domain: %s, Role: %s", domain, role)
-		outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
-		log.Debugf("Saving Role Token[%d bytes] at %s", len(rt), outPath)
-		if err := w.AddBytes(outPath, 0644, []byte(rt)); err != nil {
-			return fmt.Errorf("unable to save role token: %w", err)
-		}
+	// wait for ALL token updates to complete
+	wg.Wait()
+	log.Infof("Saved tokens to files. accesstoken:success[%d],error[%d]; roletoken:success[%d],error[%d]", int64(len(d.atTargetDomainRolesToFile))-atErrorCount.Load(), atErrorCount.Load(), int64(len(d.rtTargetDomainRolesToFile))-rtErrorCount.Load(), rtErrorCount.Load())
+
+	// collect errors
+	close(echan)
+	errs := make([]error, 0, len(d.atTargetDomainRolesToFile)+len(d.rtTargetDomainRolesToFile))
+	for err := range echan {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+// writeFileWithRetry attempts multiple times to write a file for the given token (AT or RT) by utilizing writeFile()
+func (d *tokenService) writeFileWithRetry(ctx context.Context, maxElapsedTime time.Duration, token Token, outPath string, tt mode) error {
+	operation := func() error {
+		return d.writeFile(token, outPath, tt)
+	}
+	notifyOnErr := func(err error, backoffDelay time.Duration) {
+		log.Errorf("Failed to output the token to a file: %s. Retrying in %s", err.Error(), backoffDelay)
+	}
+	return backoff.RetryNotify(operation, newExponentialBackOff(ctx, maxElapsedTime), notifyOnErr)
+}
+
+// writeFile outputs given token (AT or RT) as file
+func (d *tokenService) writeFile(token Token, outPath string, tt mode) error {
+	w := util.NewWriter()
+	tokenType := ""
+	switch tt {
+	case mACCESS_TOKEN:
+		tokenType = "Access"
+	case mROLE_TOKEN:
+		tokenType = "Role"
+	default:
+		return fmt.Errorf("invalid token type: %d", tt)
+	}
+
+	domain := token.Domain()
+	role := token.Role()
+	rawToken := token.Raw()
+	if rawToken == "" {
+		// skip placeholder token added during daemon creation
 		return nil
-	})
-	if err != nil {
-		return err
 	}
 
+	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
+	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
```

Comment:
```suggestion
        // TODO: Attempting to save token file for Domain %s, Role: %s in %s , ... outPath)
	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
```

Reply:
fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b